File: services/data/app/analysis/signals.py
"""
Combines technical indicator snapshots into a simple directional signal.
Signal values: "buy" | "sell" | "hold"
Confidence: 0.0–1.0
"""

from __future__ import annotations
from dataclasses import dataclass
from app.analysis.indicators import IndicatorSnapshot


@dataclass
class Signal:
    direction:  str    # "buy" | "sell" | "hold"
    confidence: float  # 0–1
    reasons:    list[str]


def generate(snap: IndicatorSnapshot, price: float) -> Signal:
    bull_pts = 0
    bear_pts = 0
    reasons: list[str] = []

    # ── RSI ──────────────────────────────────────────────────────────────────
    if snap.rsi14 is not None:
        if snap.rsi14 < 30:
            bull_pts += 2
            reasons.append(f"RSI oversold ({snap.rsi14:.1f})")
        elif snap.rsi14 > 70:
            bear_pts += 2
            reasons.append(f"RSI overbought ({snap.rsi14:.1f})")

    # ── MACD crossover ────────────────────────────────────────────────────────
    if snap.macd is not None and snap.signal is not None:
        if snap.macd > snap.signal:
            bull_pts += 1
            reasons.append("MACD above signal")
        else:
            bear_pts += 1
            reasons.append("MACD below signal")

    # ── Price vs Bollinger Bands ──────────────────────────────────────────────
    if snap.bb_lower is not None and snap.bb_upper is not None:
        if price < snap.bb_lower:
            bull_pts += 1
            reasons.append("Price below lower BB")
        elif price > snap.bb_upper:
            bear_pts += 1
            reasons.append("Price above upper BB")

    # ── EMA trend ─────────────────────────────────────────────────────────────
    if snap.ema50 is not None:
        if price > snap.ema50:
            bull_pts += 1
            reasons.append("Price above EMA50")
        else:
            bear_pts += 1
            reasons.append("Price below EMA50")

    total = bull_pts + bear_pts
    if total == 0:
        return Signal("hold", 0.0, ["Insufficient data"])

    confidence = round(max(bull_pts, bear_pts) / total, 2)

    if bull_pts > bear_pts:
        return Signal("buy",  confidence, reasons)
    if bear_pts > bull_pts:
        return Signal("sell", confidence, reasons)
    return Signal("hold", 0.5, reasons)

File: src/lib/signals.ts
import type { Indicator, SignalVerdict } from "@/models/coin.model";

export interface SignalResult {
  verdict: SignalVerdict;
  score: number; // −100 … +100
  reasons: { label: string; bias: "bull" | "bear" | "flat" }[];
}

export const VERDICT_META: Record<
  SignalVerdict,
  { label: string; tone: "up" | "down" | "warn" }
> = {
  strong_buy: { label: "STRONG BUY", tone: "up" },
  buy: { label: "BUY", tone: "up" },
  neutral: { label: "NEUTRAL", tone: "warn" },
  sell: { label: "SELL", tone: "down" },
  strong_sell: { label: "STRONG SELL", tone: "down" },
};

/** Rule-based score from the latest indicator snapshot + current price. */
export function computeSignal(ind: Indicator | undefined, price?: number): SignalResult {
  if (!ind) return { verdict: "neutral", score: 0, reasons: [] };

  let score = 0;
  const reasons: SignalResult["reasons"] = [];

  // RSI
  if (ind.rsi14 < 30) {
    score += 30;
    reasons.push({ label: `RSI ${ind.rsi14.toFixed(0)} · oversold`, bias: "bull" });
  } else if (ind.rsi14 > 70) {
    score -= 30;
    reasons.push({ label: `RSI ${ind.rsi14.toFixed(0)} · overbought`, bias: "bear" });
  } else {
    reasons.push({ label: `RSI ${ind.rsi14.toFixed(0)} · neutral`, bias: "flat" });
  }

  // MACD vs signal
  if (ind.macd > ind.signal) {
    score += 25;
    reasons.push({ label: "MACD above signal", bias: "bull" });
  } else {
    score -= 25;
    reasons.push({ label: "MACD below signal", bias: "bear" });
  }

  // price vs moving averages
  const ref = price ?? ind.sma20;
  if (ref > ind.ema50) {
    score += 20;
    reasons.push({ label: "Price above EMA-50", bias: "bull" });
  } else {
    score -= 20;
    reasons.push({ label: "Price below EMA-50", bias: "bear" });
  }
  if (ref > ind.sma20) {
    score += 10;
  } else {
    score -= 10;
  }

  // Bollinger position
  if (ref <= ind.bbLower) {
    score += 15;
    reasons.push({ label: "At lower Bollinger band", bias: "bull" });
  } else if (ref >= ind.bbUpper) {
    score -= 15;
    reasons.push({ label: "At upper Bollinger band", bias: "bear" });
  }

  score = Math.max(-100, Math.min(100, score));

  const verdict: SignalVerdict =
    score >= 55 ? "strong_buy"
    : score >= 20 ? "buy"
    : score <= -55 ? "strong_sell"
    : score <= -20 ? "sell"
    : "neutral";

  return { verdict, score, reasons };
}
